internal/command/artifact/junit.go

```diff
@@ -0,0 +1,263 @@
+package artifact
+
+import (
+	"bytes"
+	"encoding/xml"
+	"fmt"
+	"os"
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/terraform/internal/command/format"
+	"github.com/hashicorp/terraform/internal/configs/configload"
+	"github.com/hashicorp/terraform/internal/moduletest"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+)
+
+// TestJUnitXMLFile produces a JUnit XML file at the conclusion of a test
+// run, summarizing the outcome of the test in a form that can then be
+// interpreted by tools which render JUnit XML result reports.
+//
+// The de-facto convention for JUnit XML is for it to be emitted as a separate
+// file as a complement to human-oriented output, rather than _instead of_
+// human-oriented output. To meet that expectation the method [TestJUnitXMLFile.Save]
+// should be called at the same time as the test's view reaches its "Conclusion" event.
+// If that event isn't reached for any reason then no file should be created at
+// all, which JUnit XML-consuming tools tend to expect as an outcome of a
+// catastrophically-errored test suite.
+//
+// TestJUnitXMLFile implements the Artifact interface, which allows creation of local
+// files that contains a description of a completed test suite. It is intended only
+// for use in conjunction with a View that provides the streaming output of ongoing
+// testing events.
+
+type TestJUnitXMLFile struct {
+	filename string
+
+	// A config loader is required to access sources, which are used with diagnostics to create XML content
+	configLoader *configload.Loader
+}
+
+type Artifact interface {
+	Save(*moduletest.Suite) tfdiags.Diagnostics
+}
+
+var _ Artifact = (*TestJUnitXMLFile)(nil)
+
+// NewTestJUnitXML returns a [Test] implementation that will, when asked to
+// report "conclusion", write a JUnit XML report to the given filename.
+//
+// If the file already exists then this view will silently overwrite it at the
+// point of being asked to write a conclusion. Otherwise it will create the
+// file at that time. If creating or overwriting the file fails, a subsequent
+// call to method Err will return information about the problem.
+func NewTestJUnitXMLFile(filename string, configLoader *configload.Loader) *TestJUnitXMLFile {
+	return &TestJUnitXMLFile{
+		filename:     filename,
+		configLoader: configLoader,
+	}
+}
+
+// Save takes in a test suite, generates JUnit XML summarising the test results,
+// and saves the content to the filename specified by user
+func (v *TestJUnitXMLFile) Save(suite *moduletest.Suite) tfdiags.Diagnostics {
+	var diags tfdiags.Diagnostics
+
+	// Prepare XML content
+	sources := v.configLoader.Parser().Sources()
```

Comment:
A nil pointer error is being raised here when running the `TestTest_Runs` test

Reply:
The test case that's affected keeps changing, which suggests that there might be a race condition

Reply:
Nope, just a defect in my code - the `if runner.Artifact != nil {...}` [condition](https://github.com/hashicorp/terraform/blob/872e950f1eade3e7ff56d513b7e3b28f4d814834/internal/backend/local/test.go#L176) wasn't guarding against trying to do artifact-y things when no Artifact is configured. Fixed in [872e950](https://github.com/hashicorp/terraform/pull/36304/commits/872e950f1eade3e7ff56d513b7e3b28f4d814834)